File: LeetCode/1653-Minimum-Deletions-to-Make-String-Balanced.cpp
// 1653. Minimum Deletions to Make String Balanced
class Solution {
public:

    int getF(int l, int r, vector<int>& f) {
        if(l > r) return 0;
        int c = f[r];
        if(l > 0) c -= f[l];
        return c;
    }

    int minimumDeletions(string s) {
        int n = s.length();
        vector<int> fa(n, 0), fb(n, 0);
        for(int i = 0; i < n; i++) {
            fa[i] = (s[i] == 'a') + (i > 0 ? fa[i - 1] : 0);
            fb[i] = (s[i] == 'b') + (i > 0 ? fb[i - 1] : 0);
        }

        int ans = INT_MAX;
        for(int i = -1; i < n; i++) {
            ans = min(ans, getF(0, i, fb) + getF(i + 1, n - 1, fa));
        }
        return ans;
    }
};

File: LeetCode/1653-Minimum-Deletions-to-Make-String-Balanced.java
// 1653. Minimum Deletions to Make String Balanced

class Solution {
    public int getSum(int[] f, int l, int r) {
        int n = f.length;
        int ans = f[r];
        if (l > 0) ans -= f[l - 1];
        return ans;
    }

    public int minimumDeletions(String s) {
        int n = s.length();
        int[] fa = new int[n];
        int[] fb = new int[n];
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            if (i > 0) {
                fa[i] += fa[i - 1];
                fb[i] += fb[i - 1];
            }
            if (ch == 'a') {
                fa[i]++;
            }
            else {
                fb[i]++;
            }
        }

        int ans = Integer.MAX_VALUE;
        for (int i = 0; i <= n; i++) {
            int curr = 0;
            if (i > 0) curr += getSum(fb, 0, i - 1);
            if (i < n) curr += getSum(fa, i, n - 1);
            ans = Math.min(ans, curr);
        }

        return ans;
    }
}
